### PR TITLE
Make clippy happy

### PR DESCRIPTION
**Pull request by 0xPoe**
Try `cargo +nightly clippy --all --all-targets`.
[Remove needless borrow](https://github.com/rust-lang/rustup/commit/9d35fe07b131a3421eff984371337e0721de6394)
[Use auto deref](https://github.com/rust-lang/rustup/commit/3b18348225cf9937e22653e970313f2f2db20e97)

### CODE REVIEW

File: download/src/lib.rs [hunk 1]
```diff
@@ -47,12 +47,14 @@ fn download_with_backend(
     }
 }
 
+type DownloadCallback<'a> = Option<&'a dyn Fn(Event<'_>) -> Result<()>>;
```
Comment (rbtcollins) [inline download/src/lib.rs]: This doesn't look right as a name to me. 

```
type DownloadCallback<'a> = &'a dyn Fn(Event<'_'>) -> Result<()>;
```
might, but will that make clippy happy? If not, I think we should push back on the lint.

File: src/test.rs [hunk 2]
```diff
@@ -88,7 +89,7 @@ pub fn this_host_triple() -> String {
         // For windows, this host may be different to the target: we may be
         // building with i686 toolchain, but on an x86_64 host, so run the
         // actual detection logic and trust it.
-        let tp = Box::new(currentprocess::TestProcess::default());
+        let tp: Box<TestProcess> = Box::default();
```
Comment (rbtcollins) [inline src/test.rs]: I definitely don't like this one. The former spelling is much more what we'd see in other code that doesn't have the special case of being defaultable, and consistency helps read codebases.

File: tests/dist.rs [hunk 3]
```diff
@@ -32,11 +32,11 @@ use crate::mock::{MockComponentBuilder, MockFile, MockInstallerBuilder};
 
 const SHA256_HASH_LEN: usize = 64;
 
+// This is a test helper to edit the manifest.
+type EditFn<'a> = Option<&'a dyn Fn(&str, &mut MockChannel)>;
```
Comment (rbtcollins) [inline tests/dist.rs]: Same name critique here as above - an EditFn is the parameter type to Option, the function *may take* an edit function, or not.

**Review by rbtcollins — COMMENTED** [review]
The unneeded & is pretty shallow and trivially fine. I think the &* -> & is likewise. But the other changes are contentious for me at least. Have they been promoted to stable? If not, I think these are examples where we should raise issues in the stabilisation discussion around those lints.

**Comment by 0xPoe** [comment]
> But the other changes are contentious for me at least. Have they been promoted to stable?

https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity is stable and https://rust-lang.github.io/rust-clippy/master/#box_default is not.

I am going to close my PR. Thanks!

**Comment by rbtcollins** [comment]
Is it possible to just rebase it and pop off the two concerning patches? I really do appreciate the effort you put in and don't want to see it wasted!

**Comment by 0xPoe** [comment]
> I really do appreciate the effort you put in and don't want to see it wasted!

Thanks for your review!  💚 💙 💜 💛 ❤️

Your wording yesterday seemed a bit harsh to me. (Maybe I'm just being overly sensitive.) :( 
It made me a little sad. I'm glad you recognize my work. I do want to help make this project better.

**Comment by 0xPoe** [comment]
@rbtcollins Friendly ping~

**Review by rbtcollins — APPROVED** [review]

**Comment by rbtcollins** [comment]
Failing build is in android, not related. error: failed to run custom build command for `ring v0.16.20`

**Comment by rbtcollins** [comment]
> Your wording yesterday seemed a bit harsh to me. (Maybe I'm just being overly sensitive.) :(
> It made me a little sad. I'm glad you recognize my work. I do want to help make this project better.

I'm very sorry I made you sad. Clippy's desires are not yours, and I was not intended to point my critique at your work, rather the Clippy changes. 

In general sometimes nightly clippy lints get disabled after feedback; I guess I was hoping that you'd carry the feedback on those changes back to the clippy team so they get info about how projects are receiving them.

**Pull request merged**